### PR TITLE
Add WiFi GATT service

### PR DESCRIPTION
**Pull request by ramonhpr**
This patch adds a GATT service to set up the WiFi which the gateway will connect with.

### CODE REVIEW

File: netsetup/srv_gatt_wifi.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        logging.info('WiFi SSID: %s' % self.__convert_array_to_str(value))
+        self.value = value
+
+
+class PSWDCharacteristic(Characteristic):
+    PSWD_UUID = "a8a9e49c-aa9a-d441-9bec-817bb4900d42"
+
+    def __init__(self, bus, index, service):
+        Characteristic.__init__(self, bus, index, self.PSWD_UUID,
+                                ["encrypt-write"], service)
```
Comment (vitbaq) [inline netsetup/srv_gatt_wifi.py]: Use "write" only.
Currently, the knot app setup is not doing any security process.

**Review by vitbaq — CHANGES_REQUESTED** [review]
Remove encrypt on pswd characteristic. In this first moment, KNoT setup mobile App is not doing any security process.
Also, it's suggested adding a characteristic that passes a wifi status (e.g. if the gw is connected or not). What do you think?

**Comment by ramonhpr** [comment]
@vitbaq I agree with the characteristic status. But I will add it when I make the integration with connman, because it already has a dbus property like that.

**Review by vitbaq — CHANGES_REQUESTED** [review]
Plz, add the GATT wifi service on the BLE advertisement (line 40, ble.py). If possible, don't use the UUID value directly, use from ser_gatt_wifi.py.

File: netsetup/ble_util/characteristic.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import logging
```
Comment (jcorde) [inline netsetup/ble_util/characteristic.py:8]: About the commit message for 71ed67a87764754c7015d95d4f4031908aa49c7d

Comment (jcorde) [reply netsetup/ble_util/characteristic.py:8]: Please, update the title for something more descriptive. Something like `ble: gatt: Allow value update watch` would be more fit.

Comment (jcorde) [reply netsetup/ble_util/characteristic.py:8]: Typo: 
> the signal propertiesChanged **to** be called

Comment (jcorde) [reply netsetup/ble_util/characteristic.py:8]: Please, revert:
>  changed by the **service GATT**

File: netsetup/ble_util/characteristic.py [hunk 3]
```diff
@@ -67,8 +68,13 @@ def ReadValue(self, options):
 
     @dbus_method(GATT_CHRC_IFACE, in_signature='aya{sv}')
     def WriteValue(self, value, options):
-        logging.info('Default WriteValue called, returning error')
-        raise NotSupportedException()
+        if 'write' not in self.flags:
+            raise NotSupportedException()
+        if not value:
+            logging.error('Invalid value length')
```
Comment (jcorde) [inline netsetup/ble_util/characteristic.py]: Maybe something like `No value received` would be more appropriated

File: netsetup/ble.py [hunk 4]
```diff
@@ -34,9 +34,11 @@ def __init__(self, bus):
```
Comment (jcorde) [inline netsetup/ble.py:36]: About commit message for b207a3820d740bf30ef12fa93b5abf9ac2963b46

File: netsetup/ble.py [hunk 5]
```diff
@@ -34,9 +34,11 @@ def __init__(self, bus):
 
 
 class KnotAdvertisement(Advertisement):
-    def __init__(self, bus, index, ad_name):
+    def __init__(self, bus, index, gatt_knot, ad_name):
         Advertisement.__init__(self, bus, index, 'peripheral')
-        self.add_service_uuid("a8a9e49c-aa9a-d441-9bec-817bb4900e30")
+        self.gatt_knot = gatt_knot
+        self.add_service_uuid(gatt_knot.services[0].UUID)
+        self.current_srv_gatt = 0
```
Comment (jcorde) [inline netsetup/ble.py:43]: What is this for?

File: netsetup/ble.py (hunk 4 above)
Comment (jcorde) [reply netsetup/ble.py:36]: Please, use something like:
> Add `gatt_knot` field to store application being advertised.

File: netsetup/ble.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
 from dbus.service import method as dbus_method
```
Comment (jcorde) [inline netsetup/ble.py:13]: About message for commit 0da7f37960b281803871831b11ba9612f5642bd4

Comment (jcorde) [reply netsetup/ble.py:13]: > patch **adds** a GATT

File: netsetup/srv_gatt_wifi.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    def __init__(self, bus, index):
+        Service.__init__(self, bus, index, self.UUID, True)
+        self.add_characteristic(SSIDCharacteristic(bus, 0, self))
+
+        self.ssid_signal = self.__monitor_characteristic_changed(
+            0, self.__on_ssid_changed)
+
+    def __on_ssid_changed(self, interface, changed, invalidated):
+        self.ssid = changed['Value']
+        logging.info('SSID changed %s' % convert_array_to_str(self.ssid))
```
Comment (jcorde) [inline netsetup/srv_gatt_wifi.py]: Please, use `SSID changed *to* %s`

File: netsetup/ble.py [hunk 8]
```diff
@@ -44,6 +47,13 @@ def __init__(self, bus, index, gatt_knot, ad_name):
         self.add_local_name(ad_name)
         self.include_tx_power = True
 
+    def toogle_uuid(self):
+        qtd_srv = len(self.gatt_knot.services)
```
Comment (jcorde) [inline netsetup/ble.py]: **qtd** sorry yo no hablo portugues

File: netsetup/ble.py [hunk 9]
```diff
@@ -44,6 +47,13 @@ def __init__(self, bus, index, gatt_knot, ad_name):
...
+        self.current_srv_gatt = (self.current_srv_gatt + 1) % qtd_srv
+        uuid = self.gatt_knot.services[self.current_srv_gatt].UUID
+        self.service_uuids = []
```
Comment (jcorde) [inline netsetup/ble.py:51]: This should be called right after
```
self.ad_manager.UnregisterAdvertisement(
            self.ad_knot,
            reply_handler=lambda: None,
            error_handler=on_error)
```

File: netsetup/ble.py [hunk 10]
```diff
@@ -44,6 +47,13 @@ def __init__(self, bus, index, gatt_knot, ad_name):
         self.add_local_name(ad_name)
         self.include_tx_power = True
 
+    def toogle_uuid(self):
```
Comment (jcorde) [inline netsetup/ble.py]: as this does not remove the old UUID, I see this as a `set_next_uuid()`  or something similar

File: netsetup/__main__.py [hunk 11]
```diff
@@ -87,20 +87,16 @@ def main():
```
Comment (jcorde) [inline netsetup/__main__.py:89]: About message for commit c2fb91536142c734a511282c434c4ef14feeaa2c

Comment (jcorde) [reply netsetup/__main__.py:89]: This patch toggles the service UUIDs in advertisement packet

File: netsetup/connman.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python
```
Comment (jcorde) [inline netsetup/connman.py:1]: About message for commit cecea97f1d5f4d6e1962fa640337e0c22f656d63

Please, explain what connman does.

Also, use a more descriptive title. Something like: Add initial connection manager support.

File: netsetup/connman.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    def __get_manager_interface(self):
+        try:
+            return dbus.Interface(
+                self.bus.get_object(CONNMAN_SERVICE_NAME, '/'),
+                CONNMAN_MANAGER_INTERFACE)
+        except DBusException as err:  # ignore if connmand is not running
+            logging.error('%s: Connman is probably not running.' % err)
+            return
+
+    def __init__(self):
```
Comment (jcorde) [inline netsetup/connman.py]: Please, place `__init__` as the first function of the class

File: netsetup/connman.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+CONNMAN_MANAGER_INTERFACE = '%s.Manager' % CONNMAN_SERVICE_NAME
+
+
+class ConnmanClient(object):
+    def __get_manager_interface(self):
+        try:
+            return dbus.Interface(
+                self.bus.get_object(CONNMAN_SERVICE_NAME, '/'),
+                CONNMAN_MANAGER_INTERFACE)
+        except DBusException as err:  # ignore if connmand is not running
```
Comment (jcorde) [inline netsetup/connman.py]: This exception is not only for connman errors. Please, update the comment

File: netsetup/connman.py [hunk 15]
```diff
@@ -10,6 +10,32 @@
```
Comment (jcorde) [inline netsetup/connman.py:10]: About message for commit f77f2be6be97cfaff3c09e8e5a0d0f2f583139da

Comment (jcorde) [reply netsetup/connman.py:10]: This patch **registers** a connman agent to send the password to a wifi
network when **it** is requested to **the** user.

For more **information** about connman agent API see the following link:
https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/agent-api.txt

File: netsetup/connman.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    def Cancel(self):
+        logging.info('cancel')
+
+    @dbus.service.method(CONNMAN_AGENT_INTERFACE,
+                         in_signature='oa{sv}',
+                         out_signature='a{sv}')
+    def RequestInput(self, path, fields):
+        response = {}
+        if 'Passphrase' in fields:
+            response.update({'Passphrase': self.passphrase})
```
Comment (jcorde) [inline netsetup/connman.py:37]: Where did you define `self.passphrase`?

Comment (jcorde) [reply netsetup/connman.py:37]: Did you have to?

File: netsetup/connman.py [hunk 17]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
         self.manager = self.__get_manager_interface()
         self.agent = WifiAgent(self.bus, WIFI_AGENT_PATH)
         self.manager.RegisterAgent(WIFI_AGENT_PATH)
+
+    def __enable_wifi(self, iface):
+        properties = iface.GetProperties()
+        if not properties.get('Powered'):
+            iface.SetProperty('Powered', dbus.Boolean(True))
+            logging.info('Wifi enabled')
```
Comment (jcorde) [inline netsetup/connman.py]: You're sending the command to enable the Wi-Fi, but no verification is done. Please, update the logging to something like `'Enabling wifi'` or add a confirmation step before it.

File: netsetup/connman.py [hunk 18]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
...
+
+    def __get_wifi_technology_path(self):
+        if not self.manager:
+            logging.error('Unable to get wifi technology. Connman not running')
```
Comment (jcorde) [inline netsetup/connman.py:74]: This check should be done at the constructor.

File: netsetup/connman.py [hunk 19]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
...
+
+    def __scan_wifi(self, on_scan_completed):
+        if not self.manager:
+            logging.error('Unable to connect to wifi service')
+            return []
+        path = self.__get_wifi_technology_path()
+        services = []
+
+        if not path:
+            logging.error('Not found technology wifi')
```
Comment (jcorde) [inline netsetup/connman.py:98]: Shouldn't this check be done inside `__get_wifi_technology_path()`?

File: netsetup/connman.py [hunk 20]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
...
+            return
+
+        wifi_tech = dbus.Interface(
+            self.bus.get_object(CONNMAN_SERVICE_NAME, path),
+            CONNMAN_TECHNOLOGY_INTERFACE)
```
Comment (jcorde) [inline netsetup/connman.py]: Why not including 
```
dbus.Interface(
            self.bus.get_object(CONNMAN_SERVICE_NAME, path),
            CONNMAN_TECHNOLOGY_INTERFACE)
```
inside `__get_wifi_technology_path()` and changing its name to `__get_wifi_technology()`?

File: netsetup/connman.py [hunk 21]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
         self.manager = self.__get_manager_interface()
         self.agent = WifiAgent(self.bus, WIFI_AGENT_PATH)
         self.manager.RegisterAgent(WIFI_AGENT_PATH)
+
+    def __enable_wifi(self, iface):
```
Comment (jcorde) [inline netsetup/connman.py:66]: Why calling it `iface` here, but receiving a technology?

File: netsetup/connman.py [hunk 22]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
...
+            logging.info('Wifi enabled')
+
+    def __get_wifi_technology_path(self):
+        if not self.manager:
+            logging.error('Unable to get wifi technology. Connman not running')
+            return
+
+        for obj_path, properties in self.manager.GetTechnologies():
+                if properties.get('Type') == 'wifi':
+                        return obj_path
```
Comment (jcorde) [inline netsetup/connman.py]: Shouldn't a warning be printed in case of not finding wifi technology?

File: netsetup/connman.py [hunk 23]
```diff
@@ -54,3 +55,44 @@ def __init__(self):
...
+
+        wifi_tech = dbus.Interface(
+            self.bus.get_object(CONNMAN_SERVICE_NAME, path),
+            CONNMAN_TECHNOLOGY_INTERFACE)
+
+        self.__enable_wifi(wifi_tech)
+
+        def on_services_changed(currentServices, rmServices):
+            self.signal_match.remove()
+            on_scan_completed(currentServices)
```
Comment (jcorde) [inline netsetup/connman.py:107]: Please, add a blank line after the function definition

File: netsetup/connman.py [hunk 24]
```diff
@@ -11,6 +11,7 @@
 CONNMAN_SERVICE_NAME = 'net.connman'
```
Comment (jcorde) [inline netsetup/connman.py:11]: About message for commit 7b6690bccb858bba2858e1f80e8fbb3767671afa:

Comment (jcorde) [reply netsetup/connman.py:11]: Maybe the title could be more descriptive as `connman: Add connect to Wi-Fi support`

Comment (jcorde) [reply netsetup/connman.py:11]: >This patch adds a method to connect **to** a simple network with SSID and
password.

File: netsetup/connman.py [hunk 25]
```diff
@@ -96,3 +97,28 @@ def on_services_changed(currentServices, rmServices):
 
         wifi_tech.Scan(reply_handler=lambda: logging.info('Scan complete'),
                        error_handler=lambda err: logging.error(err))
+
+    def __get_wifi_service_path(self, name):
```
Comment (jcorde) [inline netsetup/connman.py]: Why not including `__get_wifi_service_path(self, name)` as a inner function of `__do_connect(self, path)`, and change it to `__do_connect(self, name)`?

File: netsetup/connman.py [hunk 26]
```diff
@@ -96,3 +97,28 @@ def on_services_changed(currentServices, rmServices):
...
+        if not self.manager:
+            logging.error('Unable to get wifi technology. Connman not running')
+            return
+
+        for obj_path, properties in self.manager.GetServices():
```
Comment (jcorde) [inline netsetup/connman.py:120]: Please, log in case of not finding the target obj_path

File: netsetup/connman.py [hunk 27]
```diff
@@ -96,3 +97,28 @@ def on_services_changed(currentServices, rmServices):
...
+    def __do_connect(self, path):
+        logging.info(path)
+        service = dbus.Interface(
+            self.bus.get_object(CONNMAN_SERVICE_NAME, path),
+            CONNMAN_SERVICE_INTERFACE)
+
+        service.Connect(reply_handler=lambda: logging.info('Connected'),
+                        error_handler=lambda err: logging.error(err))
+
+    def connect_wifi(self, name, password):
```
Comment (jcorde) [inline netsetup/connman.py]: This is redundant with `__do_connect()`
Shouldn't this be something like `scan_and_connect_wifi()`?

**Review by jcorde — CHANGES_REQUESTED** [review]

**Review by jcorde — APPROVED** [review]

**Pull request merged**